### DISCUSSION TITLE
getting_started/introduction/key_concepts_overview

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#65] [User-contributed notes]
# getting_started/introduction/key_concepts_overview

Every game engine revolves around abstractions you use to build your applications. In Godot, a game is a tree of nodes that you group together into scenes. You can then wire these nodes so they can...

https://docs.godotengine.org/en/stable/getting_started/introduction/key_concepts_overview.html

**Comment by eliemichel** [comment]
Very nice introduction, there is however 1 concept that I would have loved to find a proper introduction for (and I haven't yet after following the whole "getting started" section): **resources**.

I get some intuition about it, but it is unclear what they belong to: are they assets stored in their own files somewhere, like images and sounds, or are they defined within a scene file? Who owns them, etc. Just a suggestion to add a word about them in here!

**Reply by Calinou** [reply]
>  but it is unclear what they belong to: are they assets stored in their own files somewhere, like images and sounds, or are they defined within a scene file

It can be both, as the resource can be built-in to the scene or can be its own `.tres`/`.res` file. The same applies to scripts, which are external by default but can be optionally made built-in to a scene.

**Reply by eliemichel** [reply]
Thanks for the clarification! It was indeed not obvious

**Reply by APIScanner** [reply]
I'm loving the simplicity of the docs myself -- you might find https://docs.godotengine.org/en/stable/getting_started/introduction/first_look_at_the_editor.html to help. Scroll to near the bottom, or CTRL + F and look for "docks"!

**Reply by alexpalade** [reply]
I found some decent documentation about Resources at Manual→Scripting→Core features→Resources. https://docs.godotengine.org/en/stable/tutorials/scripting/resources.html

**Comment by 4lias** [comment]
Speaking from a perspective of someone who switches from Unity.
I find this naming convention a little odd and confusing. "Scene" has a fairly set in stone meaning to me, so now seeing that this engine instead of "place where the certain game segment occurs" uses it as "anything in the game" is just so weird. Couldn't they use less defined word like "object" or "entity" or something else? 
It's just a little thought of mine. Trying to change my intuition will be difficult but it's just a minor inconvenience.

**Reply by dtroitskiy** [reply]
I had the similar confusion at first, and I still think the term is not the best, but I adapted my thinking quickly either way.
The terms "object" or "entity" wouldn't be descriptive enough, I guess. Unity's "prefab" is probably the closest term, as documentation itself mentions, but maybe Godot's developers didn't want to copy it, I don't know.
And for lack of a better term, I suppose at first the reasoning was as follows: we compose a node tree into a scene, but this scene could be made reusable in the end, so why not make it possible to insert a scene into a scene. I think the main controversy comes from the habit of understanding scene as something presented on the screen (and usually occupying the entire screen) as this given moment of time. But technically, for example, a scene containing only the main game character (a character prefab in Unity's terms) is still is a scene, i.e. you can also present it on the screen standalone, although by itself it doesn't make much sense. Perhaps a more practical example could be thought out, like when one scene could be a standalone scene and also used as a prefab inserted into another scene. Maybe some UI dialog would be a good example.

**Reply by martz-rgb** [reply]
I feel like maybe a 'subtree' is more descptive concept of a scene. Not sure if it should be called like this, but it is more intuitive. Especially if you worked with trees before.

**Reply by vfridkin** [reply]
Why not branches and leaves?  Just as intuitive as a scene is a character :)

**Comment by thedenisovan** [comment]
Very interesting reading material, thanks a lot! Reading trough this docs i tought of an anlaogy between JS event listeners and godot signals and felt inner warmth :)

**Reply by dtroitskiy** [reply]
Well, the concept of this kind of event handling, or observer pattern in general, is very old, so if you're coming from the webdev background, that might be the closest association for you, but I remember it back to Visual Basic in early 2000, and also when Delphi was a thing. And it's probably much older than that.

**Reply by Calinou** [reply]
Qt signals and events are probably the closest inspiration for Godot signals (reduz himself said so :slightly_smiling_face:).

**Comment by GregBro** [comment]
I know the "Nodes and Scenes" concept is confusing for a lot of people.  If you're familiar with object oriented languages then this may help make sense:  The API (Application Programmer Interface) is the library of classes that are built into Godot and which the developer has access to.  

A Node is just a class which exist in the API.

A Scene is a class which you write yourself.

So nodes and scenes are just different names for classes.

When they say "Scenes are built up from Nodes" it actually means that your class (scene) can extend a node (Inheritance) and it may contain instances of a node (Composition).  So it's just the "is a" inheritance relationship, or the "has a" composition relationship.

**Comment by DrCoolzic** [comment]
You said
> When they say "Scenes are built up from Nodes" it actually means that your class (scene) can extend a node (Inheritance)...

Should it be
When they say "Scenes are built up from Nodes" it actually means that your class (scene) extends a node (Inheritance)...

Because as far as I understand a Scene always inherit from Node ?
